app/Http/Controllers/VolunteerController.php

```diff
@@ -60,5 +61,65 @@ public static function show($id)
         return view('volunteer.show', compact('volunteer'));
     }
 
+    /**
+     *  Volunteer set a challenge to himself
+     */
+    public function createChallenge()
+    {
+            return view('volunteer.challenge');
+
+    }
+
+
+    /**
+     *  Store the challenge in the database
+     */
+    public function storeChallenge(Request $request)
+    {
+        $this->validate($request ,
+            ['events' => 'required|numeric|min:1' ,
+             'deadline' => 'required|date_format:Y-d-m'
+            ]);
+        $user_id = Auth::user()->id;
+        $challenge = new Challenge($request->all());
+        $challenge->user_id = $user_id;
+        $user = User::findOrFail($user_id);
+        $user->challenges()->save($challenge);
+        return redirect()->action('VolunteerController@show' , [$user_id]);
+
+    }
+
+    /**
+     *  The volunteer edit a challenge
+     */
+
+    public function editChallenge($challenge_id)
+    {
+        $challenge = Challenge::findOrFail($challenge_id);
+        if($challenge->user->id == Auth::user()->id)
+            return view('volunteer.editChallenge' , compact('challenge' , 'challenge_id'));
+        return redirect('/');
+    }
+
+
+
+    /**
+     *  Store the edited challenge in the database
+     */
+    public function updateChallenge(Request $request , $challenge_id)
+    {
+        $this->validate($request , ['events' => 'required|numeric|min:1']);
+
+        $challenge = Challenge::findOrFail($challenge_id);
+        if($challenge->user->id == Auth::user()->id)
+        {
+            $challenge->update($request->all());
+            return redirect()->action('VolunteerController@show' , [Auth::user()->id]);
+        }
+        return redirect('/');
+
+    }
+
+
 
```

Comment:
when creating a function, always specify the authenticated users in the middlewares. In this case, these functions are allowed for volunteers only